chameleon/views.py

```diff
@@ -1,17 +1,106 @@
 from django.conf import settings
 from django.core.urlresolvers import reverse
 from django.contrib.auth.decorators import login_required
-from django.shortcuts import render
+from django.shortcuts import render, redirect
 from user_news.models import Outage
 from djangoRT import rtUtil
-
+from urlparse import urlparse
 import logging
+import chameleon.os_login as login
+from tas import auth as tas_auth
 from pytas.models import Project
 from webinar_registration.models import Webinar
 from django.utils import timezone
+from django.core.exceptions import SuspiciousOperation
+import sys
+from django.contrib.auth.forms import AuthenticationForm
+from django.conf import settings
+from django.http import HttpResponseRedirect
 
 logger = logging.getLogger(__name__)
 
+@login_required
+def horizon_sso_login(request):
+    next = ''
+    ## first we get the url params host, webroot, and next
+    horizon_webroot = request.GET.get('webroot') if request.GET.get('webroot') else ''
+    next = request.GET.get('next') if request.GET.get('next') else ''
+    host = request.GET.get('host') if request.GET.get('host') else ''
+
+    ## now we verify the host is valid, if not valid, log the error, send to home page...
+    valid_callback_hosts = getattr(settings, 'SSO_CALLBACK_VALID_HOSTS', [])
+    if not host or not host in valid_callback_hosts:
+        logger.error('Invalid or missing host in callback by user ' \
+            + request.user.username + ', callback host was: ' + host)
+        return HttpResponseRedirect('/')
+
+    ## get the Keystone token, and add to context 
+    ## if none available, prompt user for username/password to get one
+    context = {}
+    try:
+        context['sso_token'] = request.session['unscoped_token'].get('auth_token')
+    except:
+        return manual_ks_login(request)
+
+    protocol = getattr(settings, 'SSO_CALLBACK_PROTOCOL', 'https')
+    context['host'] = protocol + '://' + host + horizon_webroot + '/auth/ccwebsso/' + '?next=' + next
+    return render(request, 'sso/sso_callback_template.html', context)
+
+@login_required
+def manual_ks_login(request):
+    if request.method == 'POST':
+        form = KSAuthForm(request, data=request.POST)
+        user = None
+        if request.user.username == request.POST.get('username') and form.is_valid():
+            login.set_unscoped_token(request)
```

Comment:
Having to check that `set_unscoped_token` _actually set the token_ is one of the consequences of having side-effects in functions like that. A more explicit way can get around this problem:

```python
# Change 'set' to 'get'
unscoped_token = login.get_unscoped_token(request)
if unscoped_token:
    # We know login was successful!
    request.session['unscoped_token'] = unscoped_token
    return horizon_sso_login(request)
else:
    return HttpResponseRedirect('/sso/horizon/unavailable')
```

In this case extracting `get_unscoped_token` to some sort of authentication helper would also make sense. Another refactoring step might be to have a helper function to set it on the request, but this would just be so that you don't have to remember what key on the session you're using:

```python
def set_session_unscoped_token(request, token):
    request.session['unscoped_token'] = token
```